[IMPROVE] UI design for Tables and tabs component on Directory

karlprieb opened the pull request:
<!-- INSTRUCTION: Your Pull Request name should start with one of the following tags -->
<!-- [NEW] For new features -->
<!-- [FIX] For bug fixes -->
<!-- [BREAK] For pull requests including breaking changes -->

<!-- INSTRUCTION: Inform the issue number that this PR closes, or remove the line below -->

A better design for tables and tab component, applied on Directory page.

Still needs to write a documentation about these new components and how the directory feature works. We can make a blogpost. What you @RocketChat/core think?

![2018-06-07 15 47 41](https://user-images.githubusercontent.com/9200155/41119835-414ce9ec-6a6a-11e8-987c-094aee984566.gif)


<!-- INSTRUCTION: Link to a https://github.com/RocketChat/docs PR with added/updated documentation or an update to the missing/outdated documentation list, see https://rocket.chat/docs/contributing/documentation/  -->

<!-- INSTRUCTION: Tell us more about your PR with screen shots if you can -->

gdelavald commented:
There is this weird delay when toggling between users and channels, could this be a performance issue?
![](http://g.recordit.co/nqWTY8Nvmp.gif)

graywolf336 commented:
Why is the name of this pull request labeled as `[FIX]` when the description says:
> New table and tab component applied on Directory page.

ggazzo commented:
the design is new, but we didnt change nothing as feature, just fixed the bad UI, or am I wrong?

geekgonecrazy commented:
Should we instead use an improve tag?  Instead of fix or new.

vynmera commented:
@geekgonecrazy Good call, an IMPROVE tag would make the most sense here.

Furthermore, by using a tag like that, we might get more people to polish up the existing parts of RC, something quite important to good UX and code :-)

ggazzo commented:
I agree @geekgonecrazy , but I thought we had just NEW, BREAK, FIX

vynmera commented:
@ggazzo that's the issue, I think~

geekgonecrazy commented:
@ggazzo right we only have the 3.  I'm suggesting we consider one more.  That way things such as component redesign can be surfaced separately so we can make it more clear the components we improve vs the ones we fix a bug only.

Like some times we completely rework something for the purpose of improving a lot of things and it does close bugs, but over all we improved functionality. 

Just an idea.  @rodrigok what do you think?  That or should we add a tag to PR's that it's design, that way they can be found that way?

ggazzo commented on packages/rocketchat-theme/client/imports/components/tabs.css:
```diff
@@ -0,0 +1,27 @@
+}
+
+.tab {
+	margin: 0 1rem;
```
I think you should use `transition: color .3s;` here

karlprieb replied:
👍

ggazzo commented on packages/rocketchat-ui/client/components/table.js:
```diff
@@ -0,0 +1,16 @@
+Template.table.events({
+	'click tbody tr'() {
+		const onItemClick = Template.instance().data.onItemClick;
+		if (onItemClick) {
```
```javascript
return onItemClick && onItemClick(this)
```

karlprieb replied:
👍

ggazzo commented on packages/rocketchat-ui/client/components/tabs.js:
```diff
@@ -0,0 +1,18 @@
+
+Template.tabs.events({
+	'click .tab'(e) {
+		$('.tab').removeClass('active');
```
I think this way will not work for nested tabs

karlprieb replied:
how nested tabs should work? We will have nested tabs? We already have a design for that?

ggazzo replied:
we dont have design to show you now, but just to think...not for now, but is a small change

gdelavald replied:
Can we do a simple `{activeClass}` variable on the helper and change according to state?

karlprieb replied:
I created a 'ReactiveVar' to keep the state of the active tab and adde a helper to check which tab is active and add an active class to this tab.

Thank you for your suggestions :)

ggazzo commented on packages/rocketchat-ui/client/views/app/directory.js:
```diff
@@ -52,6 +52,49 @@ Template.directory.helpers({
+		};
+	},
+	onTableItemClick() {
+		const searchType = Template.instance().searchType;
```
`const { searchType } = Template.instance();`

karlprieb replied:
👍

ggazzo commented:
![image](https://user-images.githubusercontent.com/5263975/41313262-e3bd9462-6e5f-11e8-93cf-f71f4a47ae9d.png)

karlprieb commented:
All done :)

@ggazzo you already did a working infinite scroll, now I just added the loading icon while your method is getting new data and I added a debounce on the search input :)

Mobile view, new sort icons:
![localhost_3000_directory iphone 6_7_8](https://user-images.githubusercontent.com/9200155/41320888-fae32a40-6e77-11e8-81b6-ce8254010406.png)

Loading state:
![2018-06-12 19 15 23](https://user-images.githubusercontent.com/9200155/41320924-14f1ba64-6e78-11e8-9ce4-027cc5139c22.gif)

ggazzo commented:
@karlprieb sure I did the infinite scroll but is not working after your changes

look here https://rocket-chat-pr-11026.herokuapp.com/directory I created enough rooms to "scroll"

karlprieb commented:
Fixed :)

gdelavald commented:
@karlprieb @ggazzo 
There are still some weird issues going on with the list data:
![image](http://g.recordit.co/axrpmA7IM9.gif)

It's reproducible by adding a search term and then switching tabs, the same happens when switching from Users to Channels:
![image](http://g.recordit.co/52NwcyaNSq.gif)

karlprieb commented:
@gdelavald I didn't see that :) I'll make a fix!

karlprieb commented:
@gdelavald it is fixed now 👍

kaiiiiiiiii commented:
Hey @karlprieb,

two more things.

1. Everything looks a bit displaced when a channel description is given.
2. Is there a reason why the description width is still limited to 200px? Doesn't make sense on large screen IMO. Sometimes our project channels have really long descriptions (100-200 words) and it would be nice to see more than just the first 5 words :)

![screen shot 2018-06-14 at 21 36 34](https://user-images.githubusercontent.com/4711233/41434290-8632d01a-701b-11e8-830b-139f703276c8.png)

karlprieb commented:
@kaiiiiiiiii Thank you for your report :)

I already fixed the first problem.
For the second and third problems we will need some help from our design team, so I won't be able to fix that today.

@ggazzo ☝️

karlprieb commented:
Hey @kaiiiiiiiii we will remove channel description from this table layout for now. In a near future we will add a extended (with descriptions) and compact layout for this table, just like we did with the sidebar.

Thank again for your reports :)
